daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ContractKeySpec.scala

```diff
@@ -0,0 +1,430 @@
+
+      val result = suffixLenientEngine
+        .interpretCommands(
+          validating = false,
+          submitters = submitters,
+          readAs = Set.empty,
+          commands = cmds,
+          ledgerTime = now,
+          submissionTime = now,
+          seeding = InitialSeeding.TransactionSeed(txSeed),
+        )
+        .consume(_ => None, lookupPackage, lookupKey)
+      result shouldBe a[Right[_, _]]
+    }
+
+    // TEST_EVIDENCE: Semantics: contract keys should be evaluated after ensure clause
+    "be evaluated after ensure clause" in {
+      val templateId =
+        Identifier(basicTestsPkgId, "BasicTests:ComputeContractKeyAfterEnsureClause")
+      val createArg =
+        ValueRecord(
+          Some(templateId),
+          ImmArray((Some[Name]("owner"), ValueParty(alice))),
+        )
+
+      val submitters = Set(alice)
+
+      val Right(cmds) = preprocessor
+        .preprocessCommands(ImmArray(CreateCommand(templateId, createArg)))
+        .consume(_ => None, lookupPackage, lookupKey)
+
+      val result = suffixLenientEngine
+        .interpretCommands(
+          validating = false,
+          submitters = submitters,
+          readAs = Set.empty,
+          commands = cmds,
+          ledgerTime = now,
+          submissionTime = now,
+          seeding = InitialSeeding.TransactionSeed(txSeed),
+        )
+        .consume(_ => None, lookupPackage, lookupKey)
+      result shouldBe a[Left[_, _]]
+      val Left(err) = result
+      err.message should not include ("Boom")
+      err.message should include("Template precondition violated")
+    }
+
+    // TEST_EVIDENCE: Semantics: contract keys must have a non-empty set of maintainers
+    "not be create if has an empty set of maintainer" in {
+      val templateId =
+        Identifier(basicTestsPkgId, "BasicTests:NoMaintainer")
+      val createArg =
+        ValueRecord(
+          Some(templateId),
+          ImmArray((Some[Name]("sig"), ValueParty(alice))),
+        )
+
+      val submitters = Set(alice)
+
+      val Right(cmds) = preprocessor
+        .preprocessCommands(ImmArray(CreateCommand(templateId, createArg)))
+        .consume(_ => None, lookupPackage, lookupKey)
+      val result = suffixLenientEngine
+        .interpretCommands(
+          validating = false,
+          submitters = submitters,
+          readAs = Set.empty,
+          commands = cmds,
+          ledgerTime = now,
+          submissionTime = now,
+          seeding = InitialSeeding.TransactionSeed(txSeed),
+        )
+        .consume(_ => None, lookupPackage, lookupKey)
+
+      inside(result) { case Left(err) =>
+        err.message should include(
+          "Update failed due to a contract key with an empty sey of maintainers"
+        )
+      }
+    }
+
+    // Note that we provide no stability for multi key semantics so
+    // these tests serve only as an indication of the current behavior
+    // but can be changed freely.
+    "multi keys" should {
```

Comment:
I think these deserve some markers. Not for the multi-keys behavior (although no harm in already including that) but for the uck mode (we run them in both).